File: FindMissingNumber.java
/**

https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/

Given an array nums of n integers where nums[i] is in the range [1, n], 
return an array of all the integers in the range [1, n] that do not appear in nums.

Input: nums = [4,3,2,7,8,2,3,1]
Output: [5,6]



Solution:
    - its given that all the elements are positive
    - Loop over the input array and for each elemement value multiply the value at index by -1 if it is not -ve
    
     for(int i =0;  i < nums.length; i++)
     {
         int index = nums[i] -1;
         if(nums[index] > 0)
         {
             nums[index] = *(-1);
         }
         // [-4,-3,-2,-7,8,2,-3,-1]
         
         // the loop over the input array to check for the +ve value index and add it to the output
     }

Time Complexity : O(N) N = input size
Space Complexity : O(1) = input size
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
**/
class Solution 
{
    public List<Integer> findDisappearedNumbers(int[] nums) 
    {
        List<Integer> output = new ArrayList<>();
        
        if( nums == null || nums.length ==0)
            return output;
            
        for(int i=0; i< nums.length; i++)
        {
            int index = Math.abs(nums[i]) -1;
            if(nums[index] > 0)
                nums[index] = *(-1);
        }
        
        for(int i=0; i< nums.length; i++)
        {
            if(nums[i] > 0)
                output.add(i+1);
        }
        
        return output; 
    }
}

File: find_missing_number.py
# Time Complexity : O(n) because we are iterating through the array twice separately
# Space Complexity : O(1) because we are not using any extra space
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
In this code  we are using the fact that the array contains numbers from 1 to n and we are using the index 
of the array to mark the number as visited. We are iterating through the array and marking the number at the 
index as visited by multiplying it by -1. Then we are iterating through the array again and if the number is 
positive then we are appending it to the result array. We are doing this because the number at the index is 
not visited and hence it is missing.
"""

class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        if nums == None or len(nums) == 0:
            return []

        result = []

        for i in range(len(nums)):
            index = abs(nums[i]) - 1
            if nums[index] > 0:
                nums[index] *= -1

        for i in range(len(nums)):
            if nums[i] > 0:
                result.append(i+1)
            else:
                nums[i] *= -1

        return result